crates/chain/src/canonical_view.rs

```diff
@@ -0,0 +1,436 @@
+    /// # let indexer = KeychainTxOutIndex::<&str>::default();
+    /// // Get unspent outputs (UTXOs) from an indexer
+    /// for (keychain, utxo) in view.filter_unspent_outpoints(indexer.outpoints().clone()) {
+    ///     println!("{} UTXO: {} sats", keychain.0, utxo.txout.value);
+    /// }
+    /// ```
+    pub fn filter_unspent_outpoints<'v, O: Clone + 'v>(
+        &'v self,
+        outpoints: impl IntoIterator<Item = (O, OutPoint)> + 'v,
+    ) -> impl Iterator<Item = (O, FullTxOut<A>)> + 'v {
+        self.filter_outpoints(outpoints)
+            .filter(|(_, txo)| txo.spent_by.is_none())
+    }
+
+    /// Calculate the total balance of the given outpoints.
+    ///
+    /// This method computes a detailed balance breakdown for a set of outpoints, categorizing
+    /// outputs as confirmed, pending (trusted/untrusted), or immature based on their chain
+    /// position and the provided trust predicate.
+    ///
+    /// # Arguments
+    ///
+    /// * `outpoints` - Iterator of `(identifier, outpoint)` pairs to calculate balance for
+    /// * `trust_predicate` - Function that returns `true` for trusted scripts. Trusted outputs
+    ///   count toward `trusted_pending` balance, while untrusted ones count toward
+    ///   `untrusted_pending`
+    /// * `min_confirmations` - Minimum confirmations required for an output to be considered
+    ///   confirmed. Outputs with fewer confirmations are treated as pending.
+    ///
+    /// # Minimum Confirmations
+    ///
+    /// The `min_confirmations` parameter controls when outputs are considered confirmed. A
+    /// `min_confirmations` value of `0` is equivalent to `1` (require at least 1 confirmation).
+    ///
+    /// Outputs with fewer than `min_confirmations` are categorized as pending (trusted or
+    /// untrusted based on the trust predicate).
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use bdk_chain::{CanonicalView, TxGraph, local_chain::LocalChain, keychain_txout::KeychainTxOutIndex};
+    /// # use bdk_core::BlockId;
+    /// # use bitcoin::hashes::Hash;
+    /// # let tx_graph = TxGraph::<BlockId>::default();
+    /// # let chain = LocalChain::from_blocks([(0, bitcoin::BlockHash::all_zeros())].into_iter().collect()).unwrap();
+    /// # let view = CanonicalView::new(&tx_graph, &chain, chain.tip().block_id(), Default::default()).unwrap();
+    /// # let indexer = KeychainTxOutIndex::<&str>::default();
+    /// // Calculate balance with 6 confirmations, trusting all outputs
+    /// let balance = view.balance(
+    ///     indexer.outpoints().into_iter().map(|(k, op)| (k.clone(), *op)),
+    ///     |_keychain, _script| true,  // Trust all outputs
+    ///     6,  // Require 6 confirmations
+    /// );
+    /// ```
+    pub fn balance<'v, O: Clone + 'v>(
+        &'v self,
+        outpoints: impl IntoIterator<Item = (O, OutPoint)> + 'v,
+        mut trust_predicate: impl FnMut(&O, ScriptBuf) -> bool,
```

Comment:
I think `trust_predicate` is not expressive enough, for instance we're missing information about the txid. Using `FullTxOut` as the operand, we have access to more things in addition to the script pubkey.

```suggestion
        mut trust_predicate: impl FnMut(&O, &FullTxOut<A>) -> bool,
```

Reply:
Good idea. Addressed in cb72f4ae9d807b2c5376cc84297501961d36ae53

Reply:
To make it even more expressive, we can change the closure to output how many confirmations the UTXO needs to be considered "finalized".

Reply:
I'm not seeing the connection between trust predicate and additional confirmations, and so I think it'd be best for them to remain separate. FTR I'm sure there's other ways for a user to filter unspents by a target number of confs, so I'm not exactly married to this approach, but it also seems convenient as most devs are already familiar with the `balance` method.